text/0000-poll-drop-ready.md

```diff
@@ -0,0 +1,480 @@
+        // at a different location, violating the guarantees of `Pin`.
+
+        let moved_field = mem::replace(&mut self.field, T::default());
+        let pinned: Box<Pin<T>> = Box::pin(moved_field);
+    }
+}
+```
+
+However, its worth noting that a user can locally reason about whether private
+fields can be moved or accessed again, based on the APIs exposed by a type,
+and, if necessary, reconstruct a pin of those fields. [The pin documentation
+contains relevant information.][pin-docs] In other words, if your drop
+implementation is implemented correctly, you can construct a `Pin` of self
+inside your `poll_drop_ready`.
+
+## Drop order
+
+The order of calls to `poll_drop_ready` can be inferred from the rest of this
+document, but it's good to specify them explicitly.
+
+Between variables, calls to `poll_drop_ready` will occur in reverse order of
+the variable's introduction, immediately prior to the calls to that variable's
+destructor.
+
+Between fields of a type, calls to `poll_drop_ready` will occur in the textual
+order of that type's declaration, with the call for the type itself occuring
+first. This is similar to the order for calls to drop. *However*, these calls
+will occur in a loop, at the level of the type being dropped, until all calls
+to `poll_drop_ready` for that value have returned `Ready`, so they will be
+"interleaved" and concurrent in practice. The program will *not* wait for each
+field to return ready before beginning to process the subsequent field.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Like all language extensions which increase the expressiveness of the language,
+this feature will increase the complexity of Rust. In particular, this feature
+sits squarely at the intersection of async code and destructors, both areas of
+the language which involved nuanced guarantees about memory and correctness.
+Therefore, the biggest drawback of adding this feature is that it increases the
+complexity of the language.
+
+It should also be noted that it introduces silent "yield points" into async
+contexts, whereas previously all yield points were visible awaits. This is just
+an extension of the normal behavior of destructors, which already can block,
+panic, take locks, park the thread, mutate global state, and so on (as they are
+intended to do).
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+There is a big trade off in this space between guaranteeing the async
+destructor code will be called in as many places as possible and making it as
+ergonomic as possible to implement async destructors. This RFC leans heavily
+toward designing `poll_drop_ready` to be called in as many instances as
+possible, even if it makes `poll_drop_ready` less straightforward to implement.
+
+We believe this decision is well-motivated for several reasons. First, it
+pushes complexity to the implementor of the async destructor, rather than
```

Comment:
Unfortunately this approach puts a very heavy burdon on the application developer. They need to be sure that all components they choose for their application are supporting async drop. E.g. in my application I might run a function which intends to perform a database transaction. If my function is cancelled (dropped), I want to rollback the transaction, but need to wait until the rollback completes.

Now I start with a framework which supports async drop. My application works. Then later on the framework gets updated and someone uses a new future combinator. Or I switch out the framework. Then suddenly my remaining code still works, but the rollback will no longer happen - without any code directly related to the rollback having been changed.

That scenario actually horrifies me, and I don't see any other language feature in either rust or any other language which could make that happen.

Reply:
I replied to this point in the out-of-line comment, but to reiterate: this is exactly the behavior of synchronous destructors, which are not guaranteed to run, but are generally relied on by all users. Leaking data without documenting should be considered a serious bug, and beginning to leak data should be considered a breaking change. This is already our community standard around synchronous destructors, and the same process would work fine for asynchronous destructors.

Reply:
Perhaps something that is unclear is what the default behavior is of some type that *contains* other values that implemented `poll_drop_ready`. Is the default that when any type drops in an async context, it will automatically check `poll_drop_ready()` of all its fields? Or does it only do so if I remember to forward `poll_drop_ready` on every type that.

Put another way, with code:

```rust
struct Envelope<T>(T);

async fn foo() {}

async fn bar() {
    let _e = Envelope(foo());
    // does `Envelope::poll_drop_ready` forward, or did I accidentally "leak"?
}
```

Reply:
There's a specific section about this; yes, the glue is called automatically, recursively, on fields of types, just like destructors. https://github.com/withoutboats/rfcs/blob/poll-drop-ready/text/0000-poll-drop-ready.md#drop-glue-in-mempoll_drop_ready

This is the same behavior of non-asynchronous destructors.

Reply:
Oh ok, thanks! I missed that somehow.